recorder/record.py

```diff
@@ -7,60 +7,98 @@
+        return f'{self.CALL_SIGN}-{weekday}-{self.start}-{date_string}-{token}.mp3'
 
-if __name__ == "__main__": 
-    start_time = time.time()    
-    filename = get_filename()
-    logging.info(f'{filename} - starting ...') 
-    blob_writer = open_blob(filename) 
-    r = requests.get(STREAM_URL, stream=True)
-    with blob_writer as f:
-        try:
-            while True:                
-                for block in r.iter_content(10000):                
-                    f.write(block)                
-                    elapsed = time.time() - start_time 
-                    if elapsed > args.duration:                        
-                        break
-                    else: 
-                        log_progress(filename, elapsed)                      
-                logging.info(f'{filename} - complete')
-                break
-        except Exception as e:
-            logging.exception('Exception occurred')
\ No newline at end of file
+    def open_blob(self, filename):
+        storage_client = storage.Client()
+        bucket = storage_client.bucket(self.BUCKET_NAME)
+        blob = bucket.blob(filename)
+        return blob.open('wb')
+
+    def log_progress(self, filename, elapsed_time):
+        progress = min(round(elapsed_time / self.duration * 100), 100)
+        if math.fmod(progress, 5) == 0 and progress > self.last_progress:
+            logging.info(f'{filename} - {progress}% complete')
+            self.last_progress = progress
+
+    def save_streamed_response(self):
+        start_time = time.time()
+        filename = self.get_filename()
+        file = self.open_blob(filename)
+        chunk_size = 512 * 1024 # 0.5MB
+        attempts = 1
+        recording = True
+
+        while recording:
+            try:
+                logging.info(f'Attempt #{attempts}')
+                with requests.get(self.STREAM_URL, stream=True) as response:
+                    response.raise_for_status()
+                    logging.info(f'{filename} - starting ...')                       
+                    for chunk in response.iter_content(chunk_size=chunk_size):
+                        if chunk:
+                            file.write(chunk)
+                        else:
+                            logging.info('No chunk to write')
+                        
+                        elapsed_time = time.time() - start_time
+                        self.log_progress(filename, elapsed_time)         
+                        if elapsed_time >= self.duration:
+                            recording = False
+                            break
+            except Exception:                
+                logging.exception('Exception occurred')
+                attempts += 1
+                if attempts <= self.MAX_ATTEMPTS:                    
+                    time.sleep(attempts)
+                    continue
+                else:
+                    sys.exit(self.MAX_ATTEMPTS_CODE)
```

Comment:
may want to explicitly log this situation and instead exit with an integer 1 
(for more standardized handling of from the perspective of the parent application)
mind you, i'm not yet super familiar with this system in general - but just thinking about future interoperability